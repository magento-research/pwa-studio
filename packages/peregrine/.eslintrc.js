const config = {
    parser: 'babel-eslint',
    extends: ['@magento'],
    rules: {
        "no-undef": "off"
    }
};

module.exports = config;
