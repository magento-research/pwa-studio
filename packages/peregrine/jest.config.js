module.exports = {
    displayName: 'Peregrine',
    setupFiles: ['<rootDir>/scripts/shim.js', '<rootDir>/scripts/fetch-mock.js']
};
