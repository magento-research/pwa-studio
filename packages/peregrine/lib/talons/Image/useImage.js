import { useCallback, useMemo, useState } from 'react';

const imageCustomStyleProperties = {
    small: '--image-size-small',
    medium: '--image-size-medium',
    large: '--image-size-large'
};

/**
 * Returns props to render an Image component.
 *
 * @param {function} props.onError callback for error of loading image
 * @param {function} props.onLoad callback for load of image
 * @param {string} placeholder - data uri for placeholder image
 */
export const useImage = props => {
    const { onError, onLoad, resourceSizes } = props;
    const [isLoaded, setIsLoaded] = useState(false);
    const [hasError, setHasError] = useState(false);

    const handleImageLoad = useCallback(() => {
        setIsLoaded(true);

        if (typeof onLoad === 'function') {
            onLoad();
        }
    }, [onLoad]);

    const handleError = useCallback(() => {
        setHasError(true);

        if (typeof onError === 'function') {
            onError();
        }
    }, [onError]);

    // Example: { '--image-size-small': '5rem' }
    const customStyleProperties = useMemo(() => {
        const result = {};

        if (!resourceSizes) {
            return result;
        }

        for (const size in resourceSizes) {
            const styleKey = imageCustomStyleProperties[size];
            const value = resourceSizes[size];

            result[styleKey] = value;
        }

        return result;
    }, [resourceSizes]);

    return {
        customStyleProperties,
        handleError,
        handleImageLoad,
        hasError,
        isLoaded
    };
};
