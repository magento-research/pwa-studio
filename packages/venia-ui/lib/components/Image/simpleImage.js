import React from 'react';
import { string, func } from 'prop-types';

/**
 * Renders an img element directly using the supplied src.
 *
 * @param {String}  props.alt - The alt attribute for the img element.
 * @param {String}  props.className - The class name to apply to the img element.
 * @param {Func}    props.handleError - The function to call if the image fails to load.
 * @param {Func}    props.handleLoad - The function to call if the image successfully loads.
 * @param {String}  props.src - The src attribute for the img element.
 */
const SimpleImage = props => {
    const { alt, className, handleError, handleLoad, src, ...rest } = props;

    return (
        <img
            alt={alt}
            className={className}
            loading="lazy"
            onError={handleError}
            onLoad={handleLoad}
            src={src}
            {...rest}
        />
    );
};

SimpleImage.propTypes = {
    alt: string.isRequired,
    className: string,
    handleError: func,
    handleLoad: func,
    src: string.isRequired
};

export default SimpleImage;
