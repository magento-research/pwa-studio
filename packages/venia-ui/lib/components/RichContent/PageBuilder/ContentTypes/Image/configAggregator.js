import {
    getBorder,
    getCssClasses,
    getMargin,
    getPadding,
    getTextAlign
} from '../../utils';

export default node => {
    const imageNode =
        node.childNodes[0].nodeName === 'A'
            ? node.childNodes[0].childNodes
            : node.childNodes;
    const props = {
        desktopImage: imageNode[0].getAttribute('src'),
        mobileImage: imageNode[1].getAttribute('src'),
        altText: imageNode[0].getAttribute('alt'),
        title: imageNode[0].getAttribute('title'),
        openInNewTab: node.childNodes[0].getAttribute('target') === '_blank',
        ...getPadding(node),
        ...getMargin(node),
        ...getBorder(imageNode[0]),
        ...getCssClasses(node),
        ...getTextAlign(node)
    };
    if (props.desktopImage === props.mobileImage) {
        props.mobileImage = null;
    }
    if (node.childNodes[0].nodeName === 'A') {
        props.link = node.childNodes[0].getAttribute('href');
        props.linkType = node.childNodes[0].getAttribute('data-link-type');
    }
    const captionElement = node.querySelector('figcaption');
    if (captionElement) {
        props.caption = captionElement.textContent;
    }
    return props;
};
