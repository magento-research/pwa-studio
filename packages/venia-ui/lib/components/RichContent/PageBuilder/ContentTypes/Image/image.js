import React from 'react';
import { arrayOf, bool, oneOf, string } from 'prop-types';
import { Link } from '@magento/venia-drivers';
import resolveLinkProps from '../../resolveLinkProps';

const Image = props => {
    const {
        desktopImage,
        mobileImage,
        altText,
        title,
        link,
        linkType,
        openInNewTab,
        caption,
        textAlign,
        border,
        borderColor,
        borderWidth,
        borderRadius,
        marginTop,
        marginRight,
        marginBottom,
        marginLeft,
        paddingTop,
        paddingRight,
        paddingBottom,
        paddingLeft,
        cssClasses = []
    } = props;

    const figureStyles = {
        textAlign,
        marginTop,
        marginRight,
        marginBottom,
        marginLeft,
        paddingTop,
        paddingRight,
        paddingBottom,
        paddingLeft
    };
    const imageStyles = {
        border,
        borderColor,
        borderWidth,
        borderRadius
    };

    const SourceFragment = mobileImage ? (
        <source media="(max-width: 768px)" srcSet={mobileImage} />
    ) : (
        ''
    );
    const PictureFragment = (
        <>
            <picture>
                {SourceFragment}
                <img
                    src={desktopImage}
                    title={title}
                    alt={altText}
                    style={imageStyles}
                    loading="lazy"
                />
            </picture>
            {caption ? <figcaption>{caption}</figcaption> : ''}
        </>
    );

    if (typeof link === 'string') {
        const linkProps = resolveLinkProps(link, linkType);
        const LinkComponent = linkProps.to ? Link : 'a';

        return (
            <figure style={figureStyles} className={cssClasses.join(' ')}>
                <LinkComponent
                    {...linkProps}
                    {...(openInNewTab ? { target: '_blank' } : '')}
                >
                    {PictureFragment}
                </LinkComponent>
            </figure>
        );
    } else {
        return (
            <figure style={figureStyles} className={cssClasses.join(' ')}>
                {PictureFragment}
            </figure>
        );
    }
};

Image.propTypes = {
    desktopImage: string,
    mobileImage: string,
    altText: string,
    title: string,
    link: string,
    linkType: oneOf(['default', 'category', 'product', 'page']),
    openInNewTab: bool,
    caption: string,
    textAlign: string,
    border: string,
    borderColor: string,
    borderWidth: string,
    borderRadius: string,
    marginTop: string,
    marginRight: string,
    marginBottom: string,
    marginLeft: string,
    paddingTop: string,
    paddingRight: string,
    paddingBottom: string,
    cssClasses: arrayOf(string)
};

export default Image;
