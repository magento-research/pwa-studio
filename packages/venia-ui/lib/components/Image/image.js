import React, { useMemo } from 'react';
import PropTypes, {
    bool,
    func,
    number,
    oneOfType,
    shape,
    string
} from 'prop-types';
import { useImage } from '@magento/peregrine/lib/talons/Image/useImage';

import { mergeClasses } from '../../classify';
import defaultClasses from './image.css';
import PlaceholderImage from './placeholderImage';
import ResourceImage from './resourceImage';
import SimpleImage from './simpleImage';

/**
 * The Image component renders a placeholder until the image is loaded.
 *
 * @param {object}   props.classes any classes to apply to this component
 * @param {bool}     props.displayPlaceholder whether or not to display a placeholder while the image loads or if it errors on load.
 * @param {function} props.onError callback for error loading image
 * @param {function} props.onLoad callback for when image loads successfully
 * @param {string}   props.placeholder the placeholder source to display while the image loads or if it errors on load
 * @param {string}   props.resource the Magento path to the image ex: /v/d/vd12-rn_main_2.jpg
 * @param {number}   props.resourceHeight the height to request for the fallback image for browsers that don't support srcset / sizes.
 * @param {object}   props.resourceSizeBreakpoints breakpoints related to resourceSizes.
 * @param {object}   props.resourceSizes image sizes used by the browser to select the image source.
 * @param {number}   props.resourceWidth the width to request for the fallback image for browsers that don't support srcset / sizes.
 * @param {string}   props.src the source of the image, ready to use in an img element
 * @param {string}   props.type the Magento image type ("image-category" / "image-product"). Used to build the resource URL.
 */
const Image = props => {
    const {
        classes: propsClasses,
        displayPlaceholder,
        onError,
        onLoad,
        placeholder,
        resource,
        resourceHeight,
        resourceSizeBreakpoints,
        resourceSizes,
        resourceWidth,
        src,
        type,
        ...rest
    } = props;

    const talonProps = useImage({
        onError,
        onLoad,
        resourceSizes
    });

    const {
        customCSSProperties,
        handleError,
        handleImageLoad,
        hasError,
        isLoaded
    } = talonProps;

    const classes = mergeClasses(defaultClasses, propsClasses);
    const containerClass = `${classes.root} ${classes.container}`;
    const isLoadedClass = isLoaded ? classes.loaded : classes.notLoaded;
    const imageClass = `${classes.image} ${isLoadedClass}`;

    // If we have a src, use it directly. If not, assume this is a resource image.
    const actualImage = src ? (
        <SimpleImage
            className={imageClass}
            handleError={handleError}
            handleLoad={handleImageLoad}
            src={src}
            {...rest}
        />
    ) : (
        <ResourceImage
            className={imageClass}
            customCSSProperties={customCSSProperties}
            handleError={handleError}
            handleLoad={handleImageLoad}
            resource={resource}
            resourceHeight={resourceHeight}
            resourceSizeBreakpoints={resourceSizeBreakpoints}
            resourceSizes={resourceSizes}
            resourceWidth={resourceWidth}
            type={type}
            {...rest}
        />
    );

    return (
        <div className={containerClass}>
            <PlaceholderImage
                classes={classes}
                customCSSProperties={customCSSProperties}
                displayPlaceholder={displayPlaceholder}
                imageHasError={hasError}
                imageIsLoaded={isLoaded}
                resourceSizes={resourceSizes}
                src={placeholder}
                {...rest}
            />
            {actualImage}
        </div>
    );
};

const conditionallyRequiredString = (props, propName, componentName) => {
    // This component needs one of src or resource to be provided.
    if (!props.src && !props.resource) {
        return new Error(
            `Missing both 'src' and 'resource' props in ${componentName}. ${componentName} needs at least one of these to be provided.`
        );
    }

    return PropTypes.checkPropTypes(
        {
            resource: string,
            src: string
        },
        props,
        propName,
        componentName
    );
};

Image.propTypes = {
    classes: shape({
        container: string,
        loaded: string,
        notLoaded: string,
        root: string
    }),
    displayPlaceholder: bool,
    onError: func,
    onLoad: func,
    placeholder: string,
    resource: conditionallyRequiredString,
    resourceHeight: oneOfType([number, string]),
    resourceSizeBreakpoints: shape({
        small: string,
        medium: string
    }),
    resourceSizes: shape({
        small: string,
        medium: string,
        large: string
    }),
    resourceWidth: oneOfType([number, string]),
    src: conditionallyRequiredString,
    type: string
};

Image.defaultProps = {
    displayPlaceholder: true
};

export default Image;
