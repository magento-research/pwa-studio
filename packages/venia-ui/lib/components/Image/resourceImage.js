import React from 'react';
import { func, number, oneOfType, shape, string } from 'prop-types';
import { resourceUrl } from '@magento/venia-drivers';
import { useResourceImage } from '@magento/peregrine/lib/talons/Image/useResourceImage';

import { generateSrcset } from '../../util/images';

/**
 * Renders a Magento resource image.
 *
 * @param {string}   props.className the class to apply to this image.
 * @param {string}   props.resource the Magento path to the image ex: /v/d/vd12-rn_main_2.jpg
 * @param {number}   props.resourceHeight the height to request for the fallback image for browsers that don't support srcset / sizes.
 * @param {object}   props.resourceSizeBreakpoints breakpoints related to resourceSizes.
 * @param {object}   props.resourceSizes image sizes used by the browser to select the image source.
 * @param {number}   props.resourceWidth the width to request for the fallback image for browsers that don't support srcset / sizes.
 * @param {string}   props.type the Magento image type ("image-category" / "image-product"). Used to build the resource URL.
 */
const ResourceImage = props => {
    const {
        className,
        customCSSProperties,
        handleError,
        handleLoad,
        resource,
        resourceHeight,
        resourceSizeBreakpoints,
        resourceSizes,
        resourceWidth,
        type,
        ...rest
    } = props;

    const talonProps = useResourceImage({
        generateSrcset,
        resource,
        resourceHeight,
        resourceSizeBreakpoints,
        resourceSizes,
        resourceUrl,
        resourceWidth,
        type
    });

    const { sizes, src, srcSet } = talonProps;

    // Note: Attributes that are allowed to be overridden must appear before the spread of `rest`.
    return (
        // eslint-disable-next-line jsx-a11y/alt-text
        <img
            loading="lazy"
            {...rest}
            className={className}
            onError={handleError}
            onLoad={handleLoad}
            sizes={sizes}
            src={src}
            srcSet={srcSet}
            style={customCSSProperties}
        />
    );
};

ResourceImage.propTypes = {
    className: string,
    handleError: func,
    handleLoad: func,
    resource: string.isRequired,
    resourceHeight: oneOfType([number, string]),
    resourceSizeBreakpoints: shape({
        small: string,
        medium: string
    }),
    resourceSizes: shape({
        small: string,
        medium: string,
        large: string
    }).isRequired,
    resourceWidth: oneOfType([number, string]),
    type: string
};

ResourceImage.defaultProps = {
    resourceSizeBreakpoints: {},
    type: 'image-product'
};

export default ResourceImage;
