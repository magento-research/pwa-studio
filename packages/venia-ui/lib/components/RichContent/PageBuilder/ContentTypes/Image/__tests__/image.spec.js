import React from 'react';
import { createTestInstance } from '@magento/peregrine';
import Image from '../image';

jest.mock('../../../../../../classify');

test('renders a Image component', () => {
    const imageProps = {
        desktopImage: 'test-image.png'
    };
    const component = createTestInstance(<Image {...imageProps} />);

    expect(component.toJSON()).toMatchSnapshot();
});

test('renders a Image component with all props configured', () => {
    const imageProps = {
        desktopImage: 'desktop-image.png',
        mobileImage: 'mobile-image.png',
        altText: 'Alt Text',
        title: 'Title Text',
        caption: 'Example Caption',
        textAlign: 'right',
        border: 'solid',
        borderColor: 'red',
        borderWidth: '10px',
        borderRadius: '15px',
        marginTop: '10px',
        marginRight: '10px',
        marginBottom: '10px',
        marginLeft: '10px',
        paddingTop: '10px',
        paddingRight: '10px',
        paddingBottom: '10px',
        paddingLeft: '10px',
        cssClasses: ['test-class']
    };
    const component = createTestInstance(<Image {...imageProps} />);

    expect(component.toJSON()).toMatchSnapshot();
});
