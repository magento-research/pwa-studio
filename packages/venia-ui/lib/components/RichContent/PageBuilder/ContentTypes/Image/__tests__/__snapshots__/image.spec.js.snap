// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a Image component 1`] = `
<figure
  className=""
  style={
    Object {
      "marginBottom": undefined,
      "marginLeft": undefined,
      "marginRight": undefined,
      "marginTop": undefined,
      "paddingBottom": undefined,
      "paddingLeft": undefined,
      "paddingRight": undefined,
      "paddingTop": undefined,
      "textAlign": undefined,
    }
  }
>
  <picture>
    
    <img
      loading="lazy"
      src="test-image.png"
      style={
        Object {
          "border": undefined,
          "borderColor": undefined,
          "borderRadius": undefined,
          "borderWidth": undefined,
        }
      }
    />
  </picture>
  
</figure>
`;

exports[`renders a Image component with all props configured 1`] = `
<figure
  className="test-class"
  style={
    Object {
      "marginBottom": "10px",
      "marginLeft": "10px",
      "marginRight": "10px",
      "marginTop": "10px",
      "paddingBottom": "10px",
      "paddingLeft": "10px",
      "paddingRight": "10px",
      "paddingTop": "10px",
      "textAlign": "right",
    }
  }
>
  <picture>
    <source
      media="(max-width: 768px)"
      srcSet="mobile-image.png"
    />
    <img
      alt="Alt Text"
      loading="lazy"
      src="desktop-image.png"
      style={
        Object {
          "border": "solid",
          "borderColor": "red",
          "borderRadius": "15px",
          "borderWidth": "10px",
        }
      }
      title="Title Text"
    />
  </picture>
  <figcaption>
    Example Caption
  </figcaption>
</figure>
`;
